### DISCUSSION TITLE
Document level click listeners interfere with dynamically rendered Links when using React v17

### DISCUSSION THREAD

**Discussion by justincy** [vercel/next.js#19082] [Help] [answered]
# Bug report

## Describe the bug

Link fails to trigger a client-side transition under the following conditions:

- You are using React v17
- You have a document level click listener (for handling the common 'click-outside to close' behavior or dropdown menus)
- The link is conditionally rendered

## To Reproduce

Minimal repro: https://codesandbox.io/s/sweet-cori-4kr2w?file=/pages/index.js

## Expected behavior

Links always trigger a client-side transition

## System information

- OS: macOS
- Browser (if applies): Chrome
- Version of Next.js: 10.0.0
- Version of Node.js: 12.8.2

**Comment by Janpot** [comment]
Aside from the behavior described in this issue, the following works for me:
```js
  useEffect(() => {
    if (!display) return;
    const closeOnClick = () => setDisplay(false);
    document.addEventListener("click", closeOnClick);
    return () => document.removeEventListener("click", closeOnClick);
  }, [display]);
```
it doesn't require capturing events nor stopping propagation and doesn't reattach an eventhandler on every render.

**Comment by justincy** [comment]
Thanks @Janpot. Looks like just adding `[display]` as a dependency for `useEffect()` fixes it. I understand how that's useful as an optimization, but I don't understand why that solves my problem.

**Comment by justincy** [comment]
@Janpot I can't get it to work without stopping propagation because without that the close handler gets the event right after the open handler.

**Comment by Janpot** [comment, marked as answer]
Ah indeed, to keep the popup open when clicked inside, you can add `stopPropagation` on the surrounding `div` https://codesandbox.io/s/sharp-poincare-u0lnb?file=/pages/index.js
